files/__addonLocation__/package.json

```diff
@@ -58,8 +84,19 @@
     "main": "addon-main.js"
   },
   "exports": {
-    ".": "./dist/index.js",
-    "./*": "./dist/*",
+    ".": "./dist/index.js", <% if (typescript) { %>
+    "./*": {
+      "types": "./dist/*.d.ts",
+      "default": "./dist/*.js"
+    },<% } else { %>
+    "./*": "./dist/*.js",<% } %>
```

Comment:
I was looking over [the rules for `moduleResolution: 'node16'` again](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing), and I _think_ we may be able to unconditionally emit `"./*": "./dist/*.js"` here. Instead of needing to specify separate `types` and `default` keys.

Excerpt from the release post (emphasis mine):

> With TypeScript’s original Node support, it would look for a "main" field, and then look for declaration files that corresponded to that entry. **For example, if "main" pointed to `./lib/index.js`, TypeScript would look for a file called `./lib/index.d.ts`.** A package author could override this by specifying a separate field called "types" (e.g. "types": "./types/index.d.ts").
>
> The new support works similarly with [import conditions](https://nodejs.org/api/packages.html). By default, **TypeScript overlays the same rules with import conditions** – if you write an import from an ES module, it will look up the import field, and from a CommonJS module, it will look at the require field. **If it finds them, it will look for a corresponding declaration file. If you need to point to a different location for your type declarations, you can add a "types" import condition.**

If we _do_ need separate `types` and `default` keys for `./*`, then I suspect we need to do the same for the `.` => `dist/index` mapping, but hopefully we don't need it in either case 🙂

Reply:
That doesn't seem to work for me. In an addon I have a setup like that:
* [`exports`](https://github.com/simonihmig/ember-stargate/blob/master/ember-stargate/package.json#L105-L108)
* [`"moduleResolution": "node16"`](https://github.com/simonihmig/ember-stargate/blob/master/test-app/tsconfig.json#L5) for the test-app

However when I replace the exports with just `"./*": "./dist/*.js"` as suggested, I get a ton of errors (`Element implicitly has an 'any' type because expression of type '"Portal"' can't be used to index type 'Globals'.`) when running glint on the test-app. 

Feel free to give it a try by yourself: https://github.com/simonihmig/ember-stargate

Reply:
I've seen similar problems to what @simonihmig has seen: The announcement and reality seem to have drifted from each other somewhat, though as with many things `"moduleResolution": "Node16"` it may be a case of there being some *very* specific condition where it does work and many cliffs one can fall off of in the attempt to reach that condition.

Reply:
Ahh I see. It looks like maybe the implicit `.js` -> `.d.ts` lookup works for literal paths but not for `*` patterns? Sorry for the wild goose chase here.